### PR TITLE
Add dsv3 pretrain script, support flops calculation

### PR DESCRIPTION
**Pull request by guyueh1**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add deepseek v3 pretrain performance testing script.

**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by guyueh1** [comment]
@cuichenx @gdengk This is still WIP, but can you review, especially the flops calculator?

File: nemo/utils/flops_formulas.py [hunk 1]
```diff
@@ -221,3 +229,50 @@ def flux(config: FLOPSConfig):
...
+    per_input_attention_flops = 6 * (bmm1_flops + bmm2_flops) * config.layers
+
+    # linear layer flops
+    per_layer_mla_params = (
+        config.hs * config.q_lora_rank
+        + config.q_lora_rank * (config.qk_head_dim + config.qk_pos_emb_head_dim) * config.attention_heads
+    )  # Q
+    per_layer_mla_params += (
+        config.hs * config.kv_lora_rank
+        + config.kv_lora_rank * (config.qk_head_dim + config.qk_pos_emb_head_dim) * config.attention_heads
```
Comment (gdengk) [inline nemo/utils/flops_formulas.py:249]: For k pos embed, the project gemm dim is `config.qk_pos_emb_head_dim` without `config.attention_heads` since the output will be broadcast concat to the K tensor.

File: nemo/utils/flops_formulas.py [hunk 2]
```diff
@@ -221,3 +229,50 @@ def flux(config: FLOPSConfig):
     )
 
     return joint_layer_flops + single_layer_flops + other_flops
+
+
+def deepseekv3(config: FLOPSConfig):
+    """Model FLOPs for DeepSeek V3"""
+
+    # self-attention flops
+    bmm1_flops = (config.qk_head_dim + config.qk_pos_emb_head_dim) * (config.enc_seq_len**2)
```
Comment (gdengk) [inline nemo/utils/flops_formulas.py:238]: To align with the nums from Mcore team, we actually put the casualmasking factor of 0.5 into consideration.

File: nemo/lightning/pytorch/callbacks/flops_callback.py [hunk 3]
```diff
@@ -80,18 +81,32 @@ def __init__(
...
+
+        if isinstance(self.model_cfg, MLATransformerConfig):
+            config_kwargs["qk_head_dim"] = self.model_cfg.qk_head_dim
+            config_kwargs["qk_pos_emb_head_dim"] = self.model_cfg.qk_pos_emb_head_dim
+            config_kwargs["v_head_dim"] = self.model_cfg.v_head_dim
+            config_kwargs["q_lora_rank"] = self.model_cfg.q_lora_rank
+            config_kwargs["kv_lora_rank"] = self.model_cfg.kv_lora_rank
+            config_kwargs["moe_layer_freq"] = self.model_cfg.moe_layer_freq
+            config_kwargs["moe_shared_expert_intermediate_size"] = self.model_cfg.moe_shared_expert_intermediate_size
+            config_kwargs["moe_ffn_hidden_size"] = self.model_cfg.moe_ffn_hidden_size
```
Comment (cuichenx) [inline nemo/lightning/pytorch/callbacks/flops_callback.py:107]: The last three are not specific to MLA. Should we pull them out of the if block?

**Comment by guyueh1** [comment]
This PR is ready, please review @cuichenx @gdengk @erhoo82

File: scripts/performance/argument_parser.py [hunk 4]
```diff
@@ -298,6 +298,15 @@ def parse_cli_args():
     def list_of_strings(arg):
         return arg.split(',')
 
+    parser.add_argument(
+        "-rm",
+        "--recompute_modules",
+        type=list_of_strings,
+        help="Comma-separated string of modules to recompute. Defaults to None",
```
Comment (erhoo82) [inline scripts/performance/argument_parser.py:305]: `Comma-separated` should be `Slash-separated`

**Review by erhoo82 — APPROVED** [review]
LGTM
